File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    answer = 0
    
    blocks = []
    
    for b in board:
        blocks.append(list(b))
        
    while True:
        remove_set = find_all_remove(blocks, m, n)    
        
        if len(remove_set) == 0:
            break
        
        answer += len(remove_set)
        blocks = remove_blocks(blocks, m, n, remove_set)
    
    
    return answer

def find_remove(blocks, m, n, x, y):
    px = [0, 1, 1, 0]
    py = [1, 1, 0, 0]
    
    icon = blocks[x][y]
    
    result = []
    
    for dx, dy in zip(px, py):
        nx, ny = x+dx, y+dy
        
        if nx < 0 or nx >= m or ny < 0 or ny >= n:
            return []
        
        if nx >= 0 and nx < m and ny >= 0 and ny < n and blocks[nx][ny] != icon:
            return []
        
        result.append((nx, ny))
        
    return result
    

def find_all_remove(blocks, m, n):
    remove_set = set()
    
    ## 삭제할 블록들 저장
    for i in range(m):
        for j in range(n):
            if blocks[i][j] != "":
                remove_set.update(find_remove(blocks, m, n, i, j))
    
    return remove_set
    
def remove_blocks(blocks, m, n, remove_set):
    
    ## 블록 아래로 밀기
    new_block = [["" for _ in range(n)] for _ in range(m)]
    for j in range(n):
        x, y = m-1 ,j
        for i in range(m-1, -1, -1):
            if (i, j) not in remove_set:
                new_block[x][y] = blocks[i][j]
                x -= 1
    
    return new_block
    
        
    
    
    

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.*;

class Solution {
    public int solution(int m, int n, String[] board) {
        int answer = 0;
        
        char[][] map = new char[m][n];
        
        // CharArray로 변경
        for (int i=0; i<m; i++) {
            map[i] = board[i].toCharArray();
        }

        boolean changed = true;
        
        while (changed) {
            changed = false;
            
            boolean[][] remove = new boolean[m][n];

            // 2x2 같은 블록 찾기
            for (int i=0; i<m-1; i++) {
                for (int j=0; j<n-1; j++) {
                    char c = map[i][j];
                    
                    if (c != ' ' && c == map[i][j+1] && c == map[i+1][j] && c == map[i+1][j+1]) {
                        // 블록 삭제 저장
                        remove[i][j] = remove[i][j+1] = remove[i+1][j] = remove[i+1][j+1] = true;
                        changed = true;
                    }
                }
            }
            
            // 블록 삭제하고, 카운트
            for (int i=0; i<m; i++) {
                for (int j=0; j<n; j++) {
                    if (remove[i][j]) {
                        map[i][j] = ' ';
                        answer++;
                    }
                }
            }
            
            // 아래로 내리기
            for (int j=0; j<n; j++) {
                int k = m-1; // 맨 아래줄부터 시작
                for (int i=m-1; i>=0; i--) {
                    if (map[i][j] != ' ') {
                        // 블록 아래부터 채우기
                        map[k][j] = map[i][j];
                        --k;
                    }
                }
                // 위에 남은 빈칸 채우기
                for (int i=0; i<=k; i++){
                    map[i][j] = ' ';
                }
            }
        }
        return answer;
    }
}
